etna/records/models.py

```diff
@@ -1,78 +1,314 @@
 from __future__ import annotations
 
-from dataclasses import dataclass, field
-from typing import Any, Dict, Optional
+import logging
+import re
+
+from dataclasses import dataclass
+from typing import Any, Dict, Optional, Tuple, Union
 
 from django.conf import settings
 from django.http import HttpRequest
 from django.urls import reverse
+from django.utils.functional import cached_property
 
 from ..analytics.mixins import DataLayerMixin
 from ..ciim.models import APIModel
-from .transforms import transform_record_result
+from ..ciim.utils import (
+    ValueExtractionError,
+    extract,
+    format_description_markup,
+    format_link,
+)
+from .converters import IAIDConverter
+
+logger = logging.getLogger(__name__)
 
 
-@dataclass
 class Record(DataLayerMixin, APIModel):
```

Comment:
@ababic Isn't it Best Practice to calculate the data and then pass them to the view as in the `dataclass` implementation of Record. With the current implementation using property methods each template tag has to resolve to the requested value by extracting using _Ask for Forgiveness_ implementations. Although its done in memory and not querying the database is the performance benefit of this implementation is an improvement over `dataclass` when accessing data through template tags?

Reply:
And we're using _Ask for Forgiveness_ implementations here too... nothing has changed there - it's just that the work is split into separate units, so a failure in one unit doesn't kill everything else.

Reply:
The number of fields to access are same for both implementations. I thought it would be an unnecessary overhead of evaluating all the template tags to get the value, than pre-populating them in a `dataclass` without overhead to get the value. The new instance initialise for every attribute tag accessed, so I am not sure if that is cheaper than pre-populating those tag attributes only once. But as you are saying so, it must be the case.

Reply:
I would encourage you to measure the rendering speed of the detail pages before and after these changes for a comparisson

Reply:
@ababic Thanks Andy for coming along with me to show this test.